wasmstation/src/renderer.rs

```diff
@@ -37,37 +35,62 @@ const TARGET_MS_PER_FRAME: Duration = Duration::from_millis((1000.0 / TARGET_FPS
 const SCREEN_LENGTH: usize = (SCREEN_SIZE * SCREEN_SIZE) as usize;
 const TEXTURE_LENGTH: usize = SCREEN_LENGTH * 3;
 
-/// Launch a game in a SDL2 window.
-pub fn launch(mut backend: impl Backend, path: &Path, display_scale: u32) -> anyhow::Result<()> {
-    let mut save_file = path.to_path_buf();
-    save_file.set_extension("disk");
+pub struct LaunchConfig {
+    pub disk_write: Box<dyn Fn([u8; 1024]) -> Result<(), String>>,
+    pub disk_read: Box<dyn Fn() -> Result<[u8; 1024], String>>,
+    pub display_scale: u32,
+    pub title: String,
+}
+
+impl Default for LaunchConfig {
+    fn default() -> Self {
+        Self {
+            disk_write: Box::new(|_| {
+                warn!("no target set for save file");
+                Ok(())
+            }),
+            disk_read: Box::new(|| {
+                warn!("no target set for save file");
+                Ok((0..1024)
+                    .into_iter()
+                    .map(|_| 0)
+                    .collect::<Vec<u8>>()
+                    .try_into()
+                    .unwrap())
+            }),
+            display_scale: 3,
+            title: "wasmstation".to_string(),
+        }
+    }
+}
 
-    if let Ok(mut data) = fs::read(&save_file) {
-        data.resize(1024, 0);
-        backend.set_save_cache(data.try_into().unwrap());
+impl LaunchConfig {
+    pub fn from_savefile(savefile: PathBuf, display_scale: u32, title: &str) -> Self {
+        Self {
+            disk_write: disk::write(savefile.clone()),
+            disk_read: disk::read(savefile),
+            display_scale,
+            title: title.to_string(),
+        }
     }
+}
```

Comment:
@upachler

Sounds like a good addition, is this what you're looking for?

Reply:
in part, yes. What I had in mind was a `renderer::launch()` function that takes the minimum that's necessary, while the rest can all be defaulted.

Something like `renderer::launch(wasm_bytes, LaunchConfig::default())`

You could have 

```rust
let mut launch_config = LaunchConfig::default();
launch_config.display_scale = scale_from_cli_param;
...

renderer::launch(wasm_bytes, launch_config);
```

My hint was hiding away those nitty gritty details from the user of `renderer::launch()`, because most of the time they will have a file system ready to use, and therefore no need to thing about the save function. Most likely, they will all want to place the save file in `$HOME/.wasmstation/saves`, so we may well make that the default, too (but that must be platform specific; on windows it's $USER I think..).

Reply:
> What I had in mind was a renderer::launch() function that takes the minimum that's necessary, while the rest can all be defaulted.

I agree that we should make it simple, in my most recent commit I did add `LaunchConfig::default()`. Right now, `LaunchConfig::default()` has read/write closures that just logs warnings to the user about not actually saving the disk. If you want, I can make `renderer::launch` take an `Option<LaunchConfig>` so the user can provide a `None` and the function will call `config.unwrap_or_default()`.

 > Most likely, they will all want to place the save file in $HOME/.wasmstation/saves, so we may well make that the default, too (but that must be platform specific; on windows it's $USER I think..).

Sure, we could save the disk as `$HOME/.wasmstation/saves/{cart_name}.disk` or, to be compatible with wasm4, just `/path/to/cart/{cart_name}.disk`. Getting a cross platform `$HOME`/`AppData` would be pretty easy with the [`dirs`](https://crates.io/crates/dirs-next) crate, I'm not too worried about that.

The problem is that `Default::default()` explicitly does not have any function parameters so there's no way to determine what file name to save the disk as. Without a {cart_name} we can't find a place to reproducibly read/write the game disk, so the user has to pass in a name *somewhere*.

If you think that `LaunchConfig::from_savefile` is too complicated, I'd be happy to add some other helper functions over it which are a bit simpler:
 - `LaunchConfig::wasm4_compatible(cart_path: &PathBuf) -> Self` which saves the disk as `/path/to/cart/{cart_name}.disk` in the same way as wasm4-native. It could also use the same default display scale and a title "WASM-4".
 - `LaunchConfig::userwide(cart_name: &str) -> Self` which stores it at `~/.wasmstation/saves/{cart_name}.disk` with the default title "wasmstation".

Reply:
Thanks for the explanation, I see clearer now as to why you designed it like this.

However, I think there's an easy way around the problem that the cart location needs to be known for some implementations of the read/write functions.
You could simply extend these functions to receive a `SaveContext`, from which they may extract the necessary data (like the current cart's FS location, and the cart name - currently its getting a `PathBuf`, but what if you want to save data in the cloud, or load carts from there?). 
Another option would be to encapsulate `read()` and `write()` in a trait (object). This way, they can have associated data that defines their context. For cart-centric saves, you initialize this with the cart; in the default case, it doesn't need that; it simply needs to know the cart name while saving (which it could again receive from a `SaveContext`).

Reply:
I like the idea of a `read()`/`write()` trait object, that's probably where I should have started with this PR. Storing the read/write functionality in callbacks is fairly opaque and doesn't let the user know what's happening. I've found myself using little callbacks in places where traits are much better practice.

How about a trait `DiskManager` (I can't think of any other names)?

 - `struct LaunchConfig<T: DiskManager>`
   - `display_scale: u32`
   - `title: String`
   - `manager: T`

 - `trait DiskManager`
   - `fn read()`
   - `fn write()`

Then there are multiple `impl DiskManager`s:
 - `Wasm4CompatibleDisk`
 - `UserwideDisk`
 
and eventually:
 - `LocalStorageDisk` - the most likely other implementation.
 
 I'll add this today and test it out 👍🏽